### PR TITLE
Update AzureMessaging iOS to 2.0.4 and Android to 0.6

### PR DESCRIPTION
**Pull request by Redth**
Updates AzureMessaging for iOS to 2.0.4

- Switch to SDK style project
- Rework cake script to modernize it
- Remove some old bits

### CODE REVIEW

File: XPlat/AzureMessaging/Android/samples/NotificationHubsSample/NotificationHubsSample.csproj [hunk 1]
```diff
@@ -39,49 +38,47 @@
...
+  <ItemGroup>
+    <PackageReference Include="Xamarin.GooglePlayServices.Gcm" Version="71.1600.0" />
+    <PackageReference Include="Xamarin.Android.Support.v7.AppCompat" Version="28.0.0.3" />
+    <PackageReference Include="Xamarin.Android.Support.v4" Version="28.0.0.3" />
+    <PackageReference Include="Xamarin.Android.Support.Compat" Version="28.0.0.3" />
+    <PackageReference Include="Xamarin.Android.Support.Core.UI" Version="28.0.0.3" />
+    <PackageReference Include="Xamarin.Android.Support.Core.Utils" Version="28.0.0.3" />
+    <PackageReference Include="Xamarin.Android.Support.Fragment" Version="28.0.0.3" />
+    <PackageReference Include="Xamarin.Android.Support.Media.Compat" Version="28.0.0.3" />
+    <PackageReference Include="Xamarin.Azure.NotificationHubs.Android" Version="0.6.0" />
```
Comment (mattleibow) [inline XPlat/AzureMessaging/Android/samples/NotificationHubsSample/NotificationHubsSample.csproj:73]: I see you have this package reference and the project reference. Isn't this going to cause conflicts? And, if you are using project references, then you shouldn't need the nocache and co.

File: XPlat/AzureMessaging/Android/samples/NuGet.config [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+    <packageSources>
+        <add key="BuildOutput" value="../../output" />
```
Comment (mattleibow) [inline XPlat/AzureMessaging/Android/samples/NuGet.config:4]: If you are using project references, then you shouldn't need this.

File: XPlat/AzureMessaging/build.cake [hunk 3]
```diff
@@ -1,48 +1,127 @@
...
+{
+	MSBuild("./Android/source/Xamarin.Azure.NotificationHubs.Android.sln", c => {
+		c.Configuration = "Release";
+		c.Restore = true;
+		c.Targets.Clear();
+		c.Targets.Add("Rebuild");
+		c.Properties.Add("DesignTimeBuild", new [] { "false" });
+		c.BinaryLogger = new MSBuildBinaryLogSettings {
+			Enabled = true,
+			FileName = "./output/libs.binlog"
```
Comment (mattleibow) [inline XPlat/AzureMessaging/build.cake]: Maybe rename this to `libs-android` and `libs-ios` for iOS. Same with the samples and nuget variants

File: XPlat/AzureMessaging/iOS/samples/NotificationHubsSample/NotificationHubsSample.csproj [hunk 4]
```diff
@@ -101,10 +105,7 @@
     <Compile Include="HomeViewController.cs" />
   </ItemGroup>
   <ItemGroup>
-    <ProjectReference Include="..\..\source\Xamarin.Azure.NotificationHubs.iOS.csproj">
-      <Project>{99ECFF92-CA4E-4473-ADFC-5D7F3ABBC4C6}</Project>
-      <Name>Xamarin.Azure.NotificationHubs.iOS</Name>
-    </ProjectReference>
+    <PackageReference Include="Xamarin.Azure.NotificationHubs.iOS" Version="2.0.4" />
```
Comment (mattleibow) [inline XPlat/AzureMessaging/iOS/samples/NotificationHubsSample/NotificationHubsSample.csproj:108]: Same as with the Android, we shouldn't use both project references and package references. I would suggest project references as the binding project is pretty simple and does not use a targets file. Then we don't need the nocache and co.

File: XPlat/AzureMessaging/iOS/samples/NuGet.config [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+    <packageSources>
+        <add key="BuildOutput" value="../../output" />
```
Comment (mattleibow) [inline XPlat/AzureMessaging/iOS/samples/NuGet.config:4]: Not needed for project references.

File: XPlat/AzureMessaging/iOS/source/Xamarin.Azure.NotificationHubs.iOS.csproj [hunk 6]
```diff
@@ -1,59 +1,77 @@
...
+    <PackageReference Include="Xamarin.Build.Download" Version="0.4.11" />
   </ItemGroup>
-  <Import Project="$(MSBuildExtensionsPath)\Xamarin\iOS\Xamarin.iOS.ObjCBinding.CSharp.targets" />
+  
+<!-- We are embedding a framework using NativeReference tag but, on Windows with .NET Core project, 
+     this tags means another thing. On windows, if a NativeReference exists it creates a file named
+     Native.$(AssemblyName).manifest which on macOS is not being generated. This is just to satisfy
+     the MSBuild packing process as it's looking for it. It seems not to be packed in the end. -->
+  <Target Name="_TouchNativeManifestWorkaround" BeforeTargets="GenerateNuspec" Condition="!Exists('$(OutputPath)$(TargetFrameworks)\Native.$(AssemblyName).manifest')">
+    <Touch Files="$(OutputPath)$(TargetFrameworks)\Native.$(AssemblyName).manifest" AlwaysCreate="True" />
```
Comment (mattleibow) [inline XPlat/AzureMessaging/iOS/source/Xamarin.Azure.NotificationHubs.iOS.csproj]: This is not needed anymore with the msbuild extras.

File: XPlat/AzureMessaging/iOS/source/Xamarin.Azure.NotificationHubs.iOS.csproj [hunk 7]
```diff
@@ -1,59 +1,77 @@
...
-    </NativeReference>
+    <ObjcBindingCoreSource Include="StructsAndEnums.cs" Condition=" '$(EnableDefaultCompileItems)' == 'true'" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ObjcBindingApiDefinition Include="ApiDefinition.cs" Condition=" '$(EnableDefaultCompileItems)' == 'true'" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Xamarin.Build.Download" Version="0.4.11" />
```
Comment (mattleibow) [inline XPlat/AzureMessaging/iOS/source/Xamarin.Azure.NotificationHubs.iOS.csproj]: Do we need the downloader?

**Review by mattleibow — CHANGES_REQUESTED** [review]

File: XPlat/AzureMessaging/Android/samples/NotificationHubsSample/NotificationHubsSample.csproj (hunk 1 above)
Comment (Redth) [reply XPlat/AzureMessaging/Android/samples/NotificationHubsSample/NotificationHubsSample.csproj:73]: Forgot to remove project ref, gone now.

File: XPlat/AzureMessaging/Android/samples/NuGet.config (hunk 2 above)
Comment (Redth) [reply XPlat/AzureMessaging/Android/samples/NuGet.config:4]: Kept packageref, so keeping this

File: XPlat/AzureMessaging/iOS/samples/NotificationHubsSample/NotificationHubsSample.csproj (hunk 4 above)
Comment (Redth) [reply XPlat/AzureMessaging/iOS/samples/NotificationHubsSample/NotificationHubsSample.csproj:108]: Don't have both in here, and yeah we could just keep project reference, but I kind of like the fact that the sample is a real end to end test of the full package and not just a reference to the project.

**Review by mattleibow — APPROVED** [review]
LGTM

**Pull request merged**